tests/FSharp.Test.Utilities/Compiler.fs

```diff
@@ -0,0 +1,508 @@
+
+        let references = processReferences fsSource.References
+
+        let compilation = Compilation.Create(source, sourceKind, output, options, references)
+
+        compileFSharpCompilation compilation fsSource.IgnoreWarnings
+
+    and private compileCSharpCompilation (compilation: CSharpCompilation) : CompilationResult =
+
+        let outputPath = Path.Combine(Path.GetTempPath(), "FSharpCompilerTests", Path.GetRandomFileName())
+
+        Directory.CreateDirectory(outputPath) |> ignore
+
+        let filename = compilation.AssemblyName
+        let output = Path.Combine(outputPath, Path.ChangeExtension(filename, ".dll"))
+
+        let cmplResult = compilation.Emit (output)
+
+        let result =
+            { OutputPath  = None
+              Adjust      = 0
+              Warnings    = []
+              Errors      = [] }
+
+        if cmplResult.Success then
+            Success { result with OutputPath  = Some output }
+        else
+            Failure result
+
+    and private compileCSharp (csSource: CSharpCompilationSource) : CompilationResult =
+
+        let source = getSource csSource.Source
+        let name = if Option.isSome csSource.Name then csSource.Name.Value else Guid.NewGuid().ToString ()
+
+        let additionalReferences =
+            match processReferences csSource.References with
+            | [] -> ImmutableArray.Empty
+            | r  -> (List.map asMetadataReference r).ToImmutableArray().As<MetadataReference>()
+
+        let references = TargetFrameworkUtil.getReferences csSource.TargetFramework
+
+        let lv =
+          match csSource.LangVersion with
+            | CSharpLanguageVersion.CSharp8 -> LanguageVersion.CSharp8
+            | _ -> LanguageVersion.Default
+
+        let cmpl =
+          CSharpCompilation.Create(
+            name,
+            [ CSharpSyntaxTree.ParseText (source, CSharpParseOptions lv) ],
+            references.As<MetadataReference>().AddRange additionalReferences,
+            CSharpCompilationOptions (OutputKind.DynamicallyLinkedLibrary))
+
+        cmpl |> compileCSharpCompilation
+
+    and compile (cUnit: CompilationUnit) : CompilationResult =
+        match cUnit with
+        | FS fs -> compileFSharp fs
+        | CS cs -> compileCSharp cs
+        | _ -> failwith "TODO"
+
+    let private typecheckFSharpWithBaseline (options: string list) (dir: string) (file: string) : CompilationResult =
+        // Since TypecheckWithErrorsAndOptionsAgainsBaseLine throws if doesn't match expected baseline,
+        // We return a successfull CompilationResult if it succeeds.
+        CompilerAssert.TypeCheckWithErrorsAndOptionsAgainstBaseLine (Array.ofList options) dir file
+
+        Success
+            { OutputPath  = None
+              Adjust      = 0
+              Warnings    = []
+              Errors      = [] }
+
+    let private typecheckFSharpSource (fsSource: FSharpCompilationSource) : CompilationResult =
```

Comment:
Is it possible we would like have tests with multiple source files?

Reply:
Yep, it's tracked in https://github.com/dotnet/fsharp/issues/9746